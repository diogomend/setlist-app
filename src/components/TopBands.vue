<template>
  <div class="top-bands">
    <v-card-title class="deep-orange--text darken-1 text-h3"
      >Top Artists</v-card-title
    >
    <div
      v-for="(artist, i) in getTopArtists"
      :key="i"
      class="top-bands-container col-12 col-xs-6 col-sm-6 col-md-4"
    >
      <artist-card :name="artist.name" :mbid="artist.mbid" :color="colors[i]">
        <div slot="content">
          <v-layout row wrap class="text-left">
            <v-flex class="align-self-center ml-5">
              <div class="top-bands-attended">
                Attended <b>{{ artist.qty }}</b> times
              </div>
            </v-flex>
          </v-layout>
        </div>
      </artist-card>
    </div>
  </div>
</template>

<script>
import { mapGetters } from "vuex";
import ArtistCard from "@/components/ArtistCard";
export default {
  components: {
    ArtistCard
  },
  data: () => ({
    colors: [
      "#965555",
      "#96556d",
      "#965594",
      "#655596",
      "#558496",
      "#559657",
      "#967355"
    ]
  }),
  computed: {
    ...mapGetters({
      getTopArtists: "setlist/getTopArtists"
    })
  }
};
</script>

<style scoped lang="scss">
.top-bands {
  margin: 48px 0;
}
.top-bands-container {
  display: inline-block;
  padding: 0;
  margin: 12px 0;
  @media screen and (min-width: 960px) {
    padding: 16px;
    margin-bottom: 0;
  }
}

.top-bands-item {
  width: 100%;
  overflow: hidden;
  display: flex;
  background-color: tomato;
  height: 150px;
}

.top-bands-qty {
  font-size: 32px;
  display: flex;
  align-items: center;
  background: rgba(0, 0, 0, 0.7);
  color: white;
  width: 32px;
}

.top-bands-qty > * {
  display: block;
  text-align: center;
  width: 100%;
}

.top-bands-title {
  flex-grow: 1;
  text-align: left;
}
/deep/ .top-bands-attended {
  font-size: 16px;
  color: #ababab;
}

.top-bands-image {
  width: 50px;
  @media screen and (min-width: 768px) {
    width: 150px;
  }
}
</style>
