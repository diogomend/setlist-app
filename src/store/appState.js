const actions = {
  toggleDrawer: ({ commit }, toggle) => {
    commit("TOGGLE_DRAWER", toggle);
  },
  toggleLoading: ({ commit }, toggle) => {
    commit("TOGGLE_LOADING", toggle);
  }
};

const mutations = {
  TOGGLE_DRAWER(state, bool) {
    state.drawer = bool;
  },
  TOGGLE_LOADING(state, isLoading) {
    state.isLoading = isLoading;
  }
};

const getters = {
  getDrawer: state => state.drawer,
  getLoading: state => state.isLoading
};

const state = () => ({
  drawer: false,
  isLoading: true
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
