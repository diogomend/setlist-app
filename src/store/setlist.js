import setlistService from "@/service/setlist";

const actions = {
  fetchSetlists: async ({ commit, dispatch }) => {
    try {
      const { data } = await setlistService.getSetlists();
      commit("UPDATE_SETLISTS", data);
      commit("UPDATE_ARTISTS", data);
      //return Promise.resolve(data);
      return true;
    } catch (err) {
      //return Promise.reject('Could not load setlist information');
      return false;
    } finally {
      dispatch("appState/toggleLoading", false, { root: true })
    }
  }
};

const mutations = {
  UPDATE_SETLISTS(state, data) {
    state.setlists = data;
  },
  UPDATE_ARTISTS(state, data) {
    state.artists = getArtists(data);
  }
};

const getters = {
  getTotalConcerts: state => state.setlists.length,
  getShow: state => id => {
    return Object.values(state.setlists).find(element => element.id === id);
  },
  getArtist: state => mbid => {
    const data = Object.values(state.setlists).find(
      element => element.artist.mbid === mbid
    );
    if (data && Object.prototype.hasOwnProperty.call(data, "artist")) {
      return data.artist;
    }

    return null;
  },
  getArtists: state => {
    const artists = getArtists(state.setlists);

    return Object.values(artists).sort((a, b) => (a.name > b.name ? 1 : -1));
  },
  getArtistShows: state => mbid => {
    return state.setlists.reduce((ac, setlist) => {
      if (setlist.artist.mbid === mbid) {
        ac.push(setlist);
      }

      return ac;
    }, []);
  },
  getTopArtists: state => {
    const artists = Object.values(state.artists).sort((a, b) =>
      a.qty < b.qty ? 1 : -1
    );
    return artists.slice(0, 6);
  },
  getTotalArtists: state => Object.keys(state.artists).length,
  getYearlyConcerts: state => {
    return state.setlists.reduce((ac, setlist) => {
      const dateSplit = setlist.eventDate.split("-");
      const year = parseInt(dateSplit[2]);

      if (Object.prototype.hasOwnProperty.call(ac, year)) {
        const merge = [...ac[year], setlist];
        return { ...ac, [year]: merge };
      }

      return { ...ac, [year]: [setlist] };
    }, {});
  },
  getTotalCountries: state => {
    const countries = state.setlists.reduce((ac, setlist) => {
      const country = setlist.venue.city.country.code;
      const qty = Object.prototype.hasOwnProperty.call(ac, country)
        ? ac[country] + 1
        : 1;

      return { ...ac, [country]: qty };
    }, {});

    return Object.keys(countries).length;
  }
};

const state = () => ({
  setlists: [],
  isLoading: true
});

const getArtists = setlists => {
  return setlists.reduce((ac, setlist) => {
    setlist.artist.qty = Object.prototype.hasOwnProperty.call(
      ac,
      setlist.artist.mbid
    )
      ? ac[setlist.artist.mbid].qty + 1
      : 1;

    return { ...ac, [setlist.artist.mbid]: setlist.artist };
  }, {});
};

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
