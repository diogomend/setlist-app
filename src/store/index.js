import Vue from "vue";
import Vuex from "vuex";
import setlist from "@/store/setlist";
import appState from "@/store/appState";

Vue.use(Vuex);

export const storeOptions = {
  modules: {
    setlist,
    appState
  }
};

export default new Vuex.Store(storeOptions);
