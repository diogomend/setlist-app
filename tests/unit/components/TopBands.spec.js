import { shallowMount } from "@vue/test-utils";
import Vue from "vue";
import { storeOptions } from "@/store/index";
import TopBands from "@/components/TopBands";
import Vuex from "vuex";
import Vuetify from "vuetify";

Vue.use(Vuex);
Vue.use(Vuetify);
describe("TopBands.vue", () => {
  let store;
  let localVue;

  beforeEach(() => {
    jest.clearAllMocks();
    storeOptions.modules.setlist.state = {
      artists: {}
    };
    store = new Vuex.Store(storeOptions);
  });

  it("is a Vue instance", () => {
    const wrapper = shallowMount(TopBands, {
      localVue,
      store
    });

    const template = wrapper.html();
    expect(template).toMatchSnapshot();
  });
});
