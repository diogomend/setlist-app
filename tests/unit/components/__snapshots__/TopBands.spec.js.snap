// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBands.vue is a Vue instance 1`] = `
<div class="top-bands">
  <v-card-title-stub class="deep-orange darken-1">Top Artists</v-card-title-stub>
</div>
`;
